### PR TITLE
Added TintedFramePlus title card

### PR DESCRIPTION
**Pull request by khthe8th**
Added TintedFramePlus title card

### CODE REVIEW

File: KHthe8th/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+![Fringe (2008) - S01E02](https://github.com/khthe8th/TitleCardMaker-CardTypes/assets/5308389/1b20569a-77f1-48eb-8dbb-e5dcba95cbdf)
+
+## Specification
+Image shown above has template:
+
+```yaml
+templates:
+  crime:
+    library: <<library>>
+    card_type: KHthe8th/TintedFramePlusTitleCard
```
Comment (CollinHeist) [inline KHthe8th/README.md:15]: Doesn't this also use a `frame_color` of `black`?

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+
+    """Characteristics of the episode text"""
+    EPISODE_TEXT_COLOR = TITLE_COLOR
+    EPISODE_TEXT_FONT = REF_DIRECTORY / 'Galey Semi Bold.ttf'
+
+    """Whether this CardType uses season titles for archival purposes"""
+    USES_SEASON_TITLE = True
+
+    """Standard class has standard archive name"""
+    ARCHIVE_NAME = 'Tinted Frame Style'
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py:46]: You'll want to change this to "Tinted Frame Plus Style" (or something else)

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+            f'-gravity south',
+            f'-font "{self.font_file}"',
+            f'-kerning {kerning}',
+            f'-interword-spacing {interword_spacing}',
+            f'-interline-spacing {interline_spacing}',
+            f'-pointsize {font_size}',
+			# Black stroke behind title text
+            *self.black_title_commands,
+            # Add remaining sub-components
+          #  *self.title_text_commands,
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py]:
```suggestion
```

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+            # Position shadow below text
+            f'+swap',
+            f'-layers merge',
+            f'+repage \)',
+            # Overlay text and shadow onto source image
+            f'-geometry +0{vertical_shift:+}',
+            f'-composite',
+        ]
+
+
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py]:
```suggestion
```
You're not using this

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+        TopLeft = Coordinate(INSET, INSET)
+        TopRight = Coordinate(self.WIDTH - INSET, INSET + BOX_WIDTH)
+
+        # This frame is uninterrupted, draw single rectangle
+        if (self.top_element == 'omit'
+            or (self.top_element == 'index'
+                and self.hide_season_text and self.hide_episode_text)
+            or (self.top_element == 'logo'
+                and (self.logo is None or not self.logo.exists()))
+            or (self.top_element == 'title' and len(self.title_text) == 0)):
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py]:
```suggestion
                and (self.logo is None or not self.logo.exists()))):
```

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 6]
```diff
@@ -0,0 +1,711 @@
...
+            )
+            element_width /= (logo_height / 150)
+            element_width *= self.logo_size
+            margin = 25
+        # Element is title text
+        elif self.top_element == 'title':
+            element_width, _ = self.get_text_dimensions(
+                self.title_text_commands, width='max', height='max',
+            )
+            margin = 10
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py]:
```suggestion
```

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 7]
```diff
@@ -0,0 +1,711 @@
...
+        # BottomLeft = Coordinate(INSET + BOX_WIDTH, self.HEIGHT - INSET)
+        BottomRight = Coordinate(self.WIDTH - INSET, self.HEIGHT - INSET)
+
+        # This frame is uninterrupted, draw single rectangle
+        if (self.bottom_element == 'omit'
+            or (self.bottom_element == 'index'
+                and self.hide_season_text and self.hide_episode_text)
+            or (self.bottom_element == 'logo'
+                and (self.logo is None or not self.logo.exists()))
+            or (self.bottom_element == 'title' and len(self.title_text) == 0)):
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py]:
```suggestion
                and (self.logo is None or not self.logo.exists()))):
```

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 8]
```diff
@@ -0,0 +1,711 @@
...
+            )
+            element_width /= (logo_height / 150)
+            element_width *= self.logo_size
+            margin = 25
+        # Element is title
+        elif self.bottom_element == 'title':
+            element_width, _ = self.get_text_dimensions(
+                self.title_text_commands, width='max', height='max',
+            )
+            margin = 10
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py]:
```suggestion
```

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 9]
```diff
@@ -0,0 +1,711 @@
+from pathlib import Path
+from typing import Literal, Optional
+
+from modules.BaseCardType import (
+    BaseCardType, Coordinate, ImageMagickCommands, Rectangle,
+)
+
+SeriesExtra = Optional
+Element = Literal['index', 'logo', 'omit', 'title']
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py]:
```suggestion
Element = Literal['index', 'logo', 'omit']
```

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 10]
```diff
@@ -0,0 +1,711 @@
...
+            grayscale: bool = False,
+            separator: str = '-',
+            stroke_color: str = 'black',
+            episode_text_color: str = None,
+            episode_text_font: Path = EPISODE_TEXT_FONT,
+            episode_text_font_size: float = 1.0,
+            episode_text_vertical_shift: int = 0,
+            frame_color: str = None,
+            frame_width: int = BOX_WIDTH,
+            top_element: Element = 'title',
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py]:
```suggestion
            top_element: Element = 'logo',
```

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 11]
```diff
@@ -0,0 +1,711 @@
...
+
+        # Validate top, middle, and bottom elements
+        def _validate_element(element: str, middle: bool = False) -> str:
+            element = str(element).strip().lower()
+            if middle and element not in ('omit', 'logo'):
+                log.warning(f'Invalid element - must be "omit" or "logo')
+                self.valid = False
+            elif (not middle
+                and element not in ('omit', 'title', 'index', 'logo')):
+                log.warning(f'Invalid element - must be "omit", "title", '
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py]:
```suggestion
                and element not in ('omit', 'index', 'logo')):
                log.warning(f'Invalid element - must be "omit", '
```

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 12]
```diff
@@ -0,0 +1,711 @@
...
+        return (custom_episode_map
+                or episode_text_format.upper() != standard_etf)
+
+
+    def create(self) -> None:
+        """
+        Make the necessary ImageMagick and system calls to create this
+        object's defined title card.
+        """
+		        # Font customizations
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py]:
```suggestion
        # Font customizations
```

File: KHthe8th/README.md (hunk 1 above)
Comment (khthe8th) [reply KHthe8th/README.md:15]: No, it seems they have all defaulted to black somehow

File: README.md [hunk 13]
```diff
@@ -32,6 +32,7 @@ In order to contribute your own custom Card Type to this repository, follow thes
...
 | CollinHeist | `CollinHeist/BlacklistTitleCard` | <img src="https://user-images.githubusercontent.com/17693271/216839561-ec4a1c27-dcdc-4869-87dd-8d592a26aee2.jpg" height="200"/> |
+| KHthe8th | `KHthe8th/TintedFramePlus` | <img src="https://private-user-images.githubusercontent.com/5308389/287501201-1b20569a-77f1-48eb-8dbb-e5dcba95cbdf.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE1ODY1MTgsIm5iZiI6MTcwMTU4NjIxOCwicGF0aCI6Ii81MzA4Mzg5LzI4NzUwMTIwMS0xYjIwNTY5YS03N2YxLTQ4ZWItOGRiYi1lNWRjYmE5NWNiZGYuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDNUMDY1MDE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzE1YzUyY2FlODRhNWM3MGRjMzdlZjRjNWZiMGZlMDRkNzExNDAxMzUyNTljZjhiOTk4ZTNlZGU2OGIxZjIyNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.w3I_BGr6uwcCOvFJ-bRKOwFT-V7_f19Opl1_M70iysI" height="200"/> |
```
Comment (CollinHeist) [inline README.md]: Looks like this is a private image, for some reason.
```suggestion
| KHthe8th | `KHthe8th/TintedFramePlus` | <img src="https://github.com/khthe8th/TitleCardMaker-CardTypes/assets/5308389/1b20569a-77f1-48eb-8dbb-e5dcba95cbdf" height="200"/> |
```

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 14]
```diff
@@ -0,0 +1,665 @@
+from pathlib import Path
+from typing import Literal, Optional
+
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py:3]:
```suggestion

from modues.Debug import log
```

**Comment by CollinHeist** [comment]
Was it your intention to add a black stroke to the bottom index text? New:
Screenshot 2023-12-03 at 1 10 16 PM
Old:
Screenshot 2023-12-03 at 1 10 45 PM

File: KHthe8th/README.md (hunk 1 above)
Comment (CollinHeist) [reply KHthe8th/README.md:15]: I see the issue, one sec

**Comment by CollinHeist** [comment]
Okay I added a fix/comment to prevent the frame color from always being black. If you would re-generate the preview cards, I think that's basically it! 

You might want to note that the top/bottom elements cannot be `title`, and that the top element is now `logo` by default. But that's up to you.

After that (and the other outstanding comments) are fixed, I'll merge this and then I'll update it for v2 / the UI later

**Comment by khthe8th** [comment]
> Okay I added a fix/comment to prevent the frame color from always being black. If you would re-generate the preview cards, I think that's basically it!
> 
> You might want to note that the top/bottom elements cannot be `title`, and that the top element is now `logo` by default. But that's up to you.
> 
> After that (and the other outstanding comments) are fixed, I'll merge this and then I'll update it for v2 / the UI later

The top element is not logo by default now, unless a change you just made did that? In my example config if I delete the top_element: logo line the top element default is omit and there is just a gap there when I regenerate.

**Comment by khthe8th** [comment]
> Was it your intention to add a black stroke to the bottom index text? New: Screenshot 2023-12-03 at 1 10 16 PM Old: Screenshot 2023-12-03 at 1 10 45 PM

Yes this was my intention, because this is the default for StandardTitleCard

**Comment by CollinHeist** [comment]
> The top element is not logo by default now, unless a change you just made did that? In my example config if I delete the top_element: logo line the top element default is omit and there is just a gap there when I regenerate.

Yes I changed that to `logo` because the default being `title` would never do anything (you commented that out). If you want the default to stay `title` (and leave a blank space) then I'd update your preview cards so people can see what a default card looks like.

> Yes this was my intention, because this is the default for StandardTitleCard

Did you still want the frame color to be adjustable? Or always black

**Comment by khthe8th** [comment]
> > The top element is not logo by default now, unless a change you just made did that? In my example config if I delete the top_element: logo line the top element default is omit and there is just a gap there when I regenerate.
> 
> Yes I changed that to `logo` because the default being `title` would never do anything (you commented that out). If you want the default to stay `title` (and leave a blank space) then I'd update your preview cards so people can see what a default card looks like.
> 
> > Yes this was my intention, because this is the default for StandardTitleCard
> 
> Did you still want the frame color to be adjustable? Or always black

Ah okay. Yea logo by default is good, the idea was to remove title because the title is always going to be shown below anyways and you wouldn't want it two places. 

My intention was for the frame color to be adjustable, and the stroke color to default to black (but be adjustable as well)

File: KHthe8th/TintedFramePlusTitleCard.py [hunk 15]
```diff
@@ -0,0 +1,666 @@
...
+            ).draw()
+        ]
+        bottom = self._frame_bottom_commands
+
+        return [
+            # Create blank canvas
+            f'\( -size {self.TITLE_CARD_SIZE}',
+            f'xc:transparent',
+            # Draw all sets of rectangles
+            f'-fill "{self.frame_color}"',
```
Comment (CollinHeist) [inline KHthe8th/TintedFramePlusTitleCard.py:478]:
```suggestion
            f'+stroke',
            f'-fill "{self.frame_color}"',
```

**Comment by CollinHeist** [comment]
Okay then I believe all the pending changes should be it. You will probably want to update the example to have the default frame color white

**Comment by khthe8th** [comment]
> Okay then I believe all the pending changes should be it. You will probably want to update the example to have the default frame color white

okay done, I downloaded and tested the new modified file (found a typo that I resolved in last commit). Updated the readme description & image, so I think that should be it now?

**Comment by CollinHeist** [comment]
Perfect, thanks for making those changes. I'll test out the card again real quick

File: README.md [hunk 16]
```diff
@@ -32,6 +32,7 @@ In order to contribute your own custom Card Type to this repository, follow thes
 | Wdvh | `Wdvh/WhiteTextTitleOnlyLogo` | <img src="https://user-images.githubusercontent.com/17693271/178131633-23a312ac-4e5f-4d1a-87cb-60b865a51fe7.jpg" height="200"/> |
 | azuravian | `azuravian/TitleColorMatch` | <img src="https://user-images.githubusercontent.com/7379812/187586521-353ba09f-30a8-424b-bbf3-ee9036c9e638.jpg" height="200"/> |
 | CollinHeist | `CollinHeist/BlacklistTitleCard` | <img src="https://user-images.githubusercontent.com/17693271/216839561-ec4a1c27-dcdc-4869-87dd-8d592a26aee2.jpg" height="200"/> |
+| KHthe8th | `KHthe8th/TintedFramePlus` | <img src="https://github.com/khthe8th/TitleCardMaker-CardTypes/assets/5308389/1b20569a-77f1-48eb-8dbb-e5dcba95cbdf" height="200"/> |
```
Comment (CollinHeist) [inline README.md]:
```suggestion
| KHthe8th | `KHthe8th/TintedFramePlus` | <img src="https://private-user-images.githubusercontent.com/5308389/287886258-d089a1b1-7458-4eaf-ad8d-59c7f332a7c1.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NTM5MTQsIm5iZiI6MTcwMTc1MzYxNCwicGF0aCI6Ii81MzA4Mzg5LzI4Nzg4NjI1OC1kMDg5YTFiMS03NDU4LTRlYWYtYWQ4ZC01OWM3ZjMzMmE3YzEuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDVUMDUyMDE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Yjg3MTMyNDY5ODE0NGZmOGYwOGQ2ZDEzNTJkNDgzNjJlNzdiY2JlOTYzZGM2N2EyYzU3YzE0NGQ4OTA4NDliNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.d7zv7dXNZf5X10dqw7AkNcewEMyAG1FxGBYMzHpfrmk" height="200"/> |
```

**Comment by CollinHeist** [comment]
Okay I did some quick test and looks like it worked - I'll merge this now. Thanks for making this! Looks really good.

**Review by CollinHeist — APPROVED** [review]

**Pull request merged**